lib/pure/asyncdispatch.nim

```diff
@@ -1851,6 +1909,105 @@ proc send*(socket: AsyncFD, data: string,
 
   return retFuture
 
+proc close(vd: VirtualEventDispatcher) =
+  assert vd.virtualHandles.len == 0, "Cannot de-Init Virtual Event Dispacter. There are still Pending Events."
+  vd.virtualMuxHandle.unregister()
+  vd.virtualMuxHandle.close()
+  vd.virtualMuxHandle = nil
+  vd.nextVirtualHandle = 0.VirtualFD
+
+proc unregister*(ev: VirtualAsyncEvent) =
+  ## Unregisters event ``ev``.
+  doAssert(ev.vFD != InvalidVirtualFD, "Event is not registered in the queue!")
+
+  let vd = ev.vd
+  withLockIfThreads ev.eventLock:
+    ev.vd.virtualHandles.del ev.vFD
+    ev.vd = nil
+    ev.cb = nil
+    ev.vFD = InvalidVirtualFD
+
+  if vd.virtualHandles.len == 0:
+    # lazy close the physical event with the Dispatcher
+    # The main reason we need this is to make
+    # hasPendingOperations still work
+    # without modifying the ioselector code.
+    vd.close
+
+proc initVirtualEventDispatcher(p: PDispatcher) =
+  p.vd.virtualMuxHandle = newAsyncEvent()
+
+  proc virtualMuxCB(fd: AsyncFD): bool {.closure,gcsafe.} =
+    # find the virtual events that triggered the physical event and
+    # add them to the callback list.
+    # Not very efficient, but requires the least coordination between threads.
+    
+    proc vcbFactory(vfd: VirtualFD) : proc() {.closure, gcsafe.} =
```

Comment:
Nit:

```suggestion
    proc vcbFactory(vfd: VirtualFD): proc() {.closure, gcsafe.} =
```

But also, I would probably try to avoid this being a closure (and use getGlobalDispatcher if possible like suggested below).

Reply:
as I mention here: https://github.com/nim-lang/Nim/pull/12372#discussion_r335726313
it's not thread-safe or consistent with the API to use `getGlobalDispatcher`.
A closure is the easiest way to get access to the dispatcher given those constraints.

As a general note:
`getGlobalDispatcher` is not appropriate for a threaded environment. You cannot assume there is only a single `asyncdispatcher` in a threaded application. IMO this might cause a lot of problems in the future.